source/slang/slang-ir-lower-error-handling.cpp

```diff
@@ -0,0 +1,239 @@
+// slang-ir-lower-error-handling.cpp
+
+#include "slang-ir-lower-error-handling.h"
+
+#include "slang-ir.h"
+#include "slang-ir-insts.h"
+
+namespace Slang
+{
+
+struct ErrorHandlingLoweringContext
+{
+    IRModule* module;
+    DiagnosticSink* diagnosticSink;
+
+    SharedIRBuilder sharedBuilder;
+
+    List<IRInst*> workList;
+    HashSet<IRInst*> workListSet;
+
+    void addToWorkList(IRInst* inst)
+    {
+        if (workListSet.Contains(inst))
+            return;
+
+        workList.add(inst);
+        workListSet.Add(inst);
+    }
+
+    void processFuncType(IRFuncType* funcType)
+    {
+        auto throwAttr = funcType->findAttr<IRFuncThrowTypeAttr>();
+        if (!throwAttr)
+            return;
+        IRBuilder builder(sharedBuilder);
+        builder.setInsertBefore(funcType);
+        auto resultType =
+            builder.getResultType(funcType->getResultType(), throwAttr->getErrorType());
+        List<IRType*> paramTypes;
+        for (UInt i = 0; i < funcType->getParamCount(); i++)
+        {
+            if (as<IRAttr>(funcType->getParamType(i)))
+                break;
+            paramTypes.add(funcType->getParamType(i));
+        }
+        auto newFuncType = builder.getFuncType(paramTypes, resultType);
+        sharedBuilder.replaceGlobalInst(funcType, newFuncType);
+    }
+
+    void processTryCall(IRTryCall* tryCall)
+    {
+        // If we see:
+        // ```
+        //      value = tryCall(callee, errorVar, successBlock, failBlock, args) : returnType
+        //  successBlock:
+        //      ... (uses value) ...
+        //  failBlock:
+        //      errorVal = load(errorVar);
+        //      (uses errorVal)
```

Comment:
This can work for now, but it isn't really an ideal encoding because we end up needing a `var` instruction for the error, and the `value` ends up visible to all blocks dominated by the `tryCall` even if they are on the failure path.

I'd much rather have this code block look like:

```
        //  tryCall(successBlock, failBlock, callee args)
        //
        //  successBlock(value : returnType):
        //      ... (uses value) ...
        //
        //  failBlock(errorVal : errorType):
        //      ... (uses errorVal) ...
```

With that encoding, we never need any `var`s and the result value/error are only visible (per SSA rules) strictly along the control-flow path where they are well-defined.